src/urlpattern_pattern_parser.cpp

```diff
@@ -0,0 +1,339 @@
+  }
+  // 10. Let type be "regexp".
+  auto type = PART_TYPE::REGEXP;
+  // Next, we convert regexp value into a part type. We make sure to go to a
+  // regular expression first so that an equivalent "regexp" token will be
+  // treated the same as a "name" or "asterisk" token.
+
+  // 11. If regexp value is parser’s segment wildcard regexp:
+  if (regexp_value == segment_wildcard_regexp) {
+    // 1. Set type to "segment-wildcard".
+    type = PART_TYPE::SEGMENT_WILDCARD;
+    // 2. Set regexp value to the empty string.
+    regexp_value.clear();
+  }
+  // 12. Else if regexp value is the full wildcard regexp value:
+  else if (regexp_value == full_wildcard_regexp_value) {
+    // 1. Set type to "full-wildcard".
+    type = PART_TYPE::FULL_WILDCARD;
+    // 2. Set regexp value to the empty string.
+    regexp_value.clear();
+  }
+  // 13. Let name be the empty string.
+  std::u32string name{};
+  // Next, we determine the part name. This can be explicitly provided by a
+  // "name" token or be automatically assigned.
+
+  // 14. If name token is not null, then set name to name token’s value.
+  if (name_token.has_value()) {
+    name = name_token.value()->value;
+  }
+  // 15. Else if regexp or wildcard token is not null:
+  else if (regexp_or_wildcard_token.has_value()) {
+    // 1. Set name to parser’s next numeric name.
+    // TODO: Needs a huge improvement
+    std::stringstream ss{};
+    ss << next_numeric_name;
+    for (char c : ss.str()) {
+      name += c;
+    }
+    // 2. Increment parser’s next numeric name by 1.
+    ++next_numeric_name;
+  }
+  // 16 If the result of running is a duplicate name given parser and name is
+  // true, then throw a TypeError.
+  if (is_duplicate_name(name)) {
+    throw std::bad_typeid();
+  }
+}
+
+// https://wicg.github.io/urlpattern/#try-to-consume-a-modifier-tokenssss
+ada_really_inline std::optional<token *>
+pattern_parser::try_to_consume_modifier_token() {
+  // 1. Let token be the result of running try to consume a token given parser
+  // and "other-modifier".
+  std::optional<token *> t = try_to_consume_token(TOKEN_TYPE::OTHER_MODIFIER);
+
+  // 2. If token is not null, then return token.
+  if (t.has_value()) {
+    return t;
+  }
+
+  // 3. Set token to the result of running try to consume a token given parser
+  // and "asterisk".
+  t = try_to_consume_token(TOKEN_TYPE::ASTERISK);
+  return t;
+}
+
+ada_really_inline std::optional<token *>
+pattern_parser::try_to_consume_regexp_or_wildcard_token(
+    std::optional<token *> &name_token) {
+  // Let token be the result of running try to consume a token given parser and
+  // "regexp".
+  std::optional<token *> regexp_or_wildcard =
+      try_to_consume_token(TOKEN_TYPE::REGEXP);
+
+  // If name token is null and token is null, then set token to the result of
+  // running try to consume a token given parser and "asterisk".
+  if (!name_token.has_value() && !regexp_or_wildcard.has_value()) {
```

Comment:
There is fast path here. If name_token does not have a value, you dont need to try to consume token right?